packages/SwingSet/src/kernel/state/kernelKeeper.js

```diff
@@ -1486,7 +1504,7 @@ export default function makeKernelKeeper(
 
   function processRefcounts() {
     if (enableKernelGC) {
-      const actions = getGCActions(); // local cache
+      const actions = new Set();
       // TODO (else buggy): change the iteration to handle krefs that get
       // added multiple times (while we're iterating), because we might do
       // different work the second time around. Something like an ordered
```

Comment:
> I'm not sure if it's describing a current risk or a refactoring gotcha

I think it's not a current risk because we don't happen to do anything in this loop that might push more things onto `maybeFreeKrefs`. I think it's a future risk, because #2069 auxdata would cause kernel objects to hold references to other krefs, and the `deleteKernelObject()` at the end would free those, which would modify `maybeFreeKrefs`.

I don't think the Set iterator property you describe is sufficient to handle this. Suppose `ko2` contains auxdata which references `ko1`. Something happens to put both `ko1` and `ko2` in `maybeFreeKrefs`, but `ko1` is a "break before make" situation, and its `reachable` count is back up to 1 (via `ko2`) by the time we reach `processRefcounts()`.

We'd iterate through the set, see `ko1` the first time, see the non-zero `reachable` count, and do nothing. Then the iterator gives us `ko2`, we delete it, which frees the auxdata, which drops the `ko1` reachable count to 0 (for good this time) and pushes `ko1` back onto `maybeFreeKrefs`.

We need that iterator to examine `ko1` again. If the iterator doesn't, we'll leak an object, because the refcount is 0 but we'll never look at it again.

(It's true that I wasn't confident of the behavior of the Set iterator when I wrote the comment, but I think I'd heard enough about it's guarantees that I was still concerned about the add-to-existing-set case).

I'll update the comment to explain the double-examination requirement, and that it shouldn't be triggered until /unless we implement #2069, or something else which allows "reentrant" release of references.

Reply:
Hrm, actually, it *is* a current risk, because promise resolutions hold data, so we could hit the above scenario even without auxdata. It would take some doing to get both krefs into `maybeFreeKrefs` at the same time.. the unit test for this will be a challenge to construct.

Also, it's a bit dubious for this code to assume that `kp.data.slots` even exists.. this works because nearly all promises are in c-lists, and we don't remove kpids from clists until they're settled, so the main way to hit `reachable === 0` for a k*p*id is to resolve or reject it, which will give it data of some sort. I think I could violate this by using `queueToKref()` to create a message with a result promise kpid  that is not yet in anybody's c-list, then put *that* kpid in the arguments of a second message, then arrange for the second message to go splat before the first message gets delivered (maybe by sending the first to a promise instead of an object). Probably not something vat code could trigger on its own, since promises are usually created by vats, which means they'll be in a c-list, which will hold the refcount firm until the promise settles.

Reply:
#9888 filed for the Set iterator issue, and #9889 for the `kp.data.slots` issue